### PR TITLE
feat: adding fee middleware support to ics27 interchain accounts

### PR DESCRIPTION
**Pull request by damiannolan**
## Description

The following PR adds support to `apps/27-interchain-accounts` to work in an application stack composed with `apps/29-fee`.

- Adding additional `version` arg to `RegisterInterchainAccount` and updating godocs.
- Removing the construction of ics27 metadata within `RegisterInterchainAccount` and passing through the new `version` arg to `MsgChannelOpenInit` directly - this supports both fee-disabled and fee-enabled channels in an application stack composed with ics29 middleware.
- Updating `simapp` to include `29-fee` in interchain accounts controller and host app stacks.
- Updating test cases to adhere to changes
- Adding migration docs

closes: #1356 

---

Before we can merge this PR, please make sure that all the following items have been
checked off. If any of the checklist items are not applicable, please leave them but
write a little note why.

- [x] Targeted PR against correct branch (see [CONTRIBUTING.md](https://github.com/cosmos/ibc-go/blob/master/CONTRIBUTING.md#pr-targeting))
- [x] Linked to Github issue with discussion and accepted design OR link to spec that describes this work.
- [x] Code follows the [module structure standards](https://github.com/cosmos/cosmos-sdk/blob/master/docs/building-modules/structure.md).
- [x] Wrote unit and integration [tests](https://github.com/cosmos/ibc-go/blob/master/CONTRIBUTING.md#testing)
- [x] Updated relevant documentation (`docs/`) or specification (`x/<module>/spec/`)
- [x] Added relevant `godoc` [comments](https://blog.golang.org/godoc-documenting-go-code).
- [x] Added a relevant changelog entry to the `Unreleased` section in `CHANGELOG.md`
- [x] Re-reviewed `Files changed` in the Github PR explorer
- [ ] Review `Codecov Report` in the comment section below once CI passes

### CODE REVIEW

File: docs/migrations/v3-to-v4.md [hunk 1]
```diff
@@ -18,14 +18,72 @@ No genesis or in-place migrations required when upgrading from v1 or v2 of ibc-g
...
 The `OnChanOpenInit` application callback has been modified.
 The return signature now includes the application version as detailed in the latest IBC [spec changes](https://github.com/cosmos/ibc/pull/629).
 
+### ICS27 - Interchain Accounts
+
+The `RegisterInterchainAccount` API has been modified to include an additional `version` argument. This change has been made in order to support ICS29 fee middelware, for relayer incentivization of ICS27 packets.
+Consumers of the `RegisterInterchainAccount` are now expected to build the appropriate JSON encoded version string themselves and pass it accordingly. 
+This should be constructed within the interchain accounts authentication module which leverages the APIs exposed via the interchain accounts `controllerKeeper`. 
+
+The following code snippet illustrates how to construct an appropriate interchain accounts Metadata and encode it as a JSON bytestring:
```
Comment (crodriguezvega) [inline docs/migrations/v3-to-v4.md:35]:
```suggestion
The following code snippet illustrates how to construct an appropriate interchain accounts `Metadata` and encode it as a JSON bytestring:
```

File: docs/migrations/v3-to-v4.md [hunk 2]
```diff
@@ -18,14 +18,72 @@ No genesis or in-place migrations required when upgrading from v1 or v2 of ibc-g
...
+    Version:                icatypes.Version,
+    ControllerConnectionId: controllerConnectionID,
+    HostConnectionId:       hostConnectionID,
+    Encoding:               icatypes.EncodingProtobuf,
+    TxType:                 icatypes.TxTypeSDKMultiMsg,
+}
+
+appVersion, err := icatypes.ModuleCdc.MarshalJSON(&icaMetadata)
+if err != nil {
+    retutn err
```
Comment (crodriguezvega) [inline docs/migrations/v3-to-v4.md:48]:
```suggestion
    return err
```

File: docs/migrations/v3-to-v4.md [hunk 3]
```diff
@@ -18,14 +18,72 @@ No genesis or in-place migrations required when upgrading from v1 or v2 of ibc-g
...
+}
+
+if err := k.icaControllerKeeper.RegisterInterchainAccount(ctx, msg.ConnectionId, msg.Owner, string(appVersion)); err != nil {
+    return err
+}
+```
+
+Similarly, if the application stack is configured to route through ICS29 fee middleware and a fee enabled channel is desired, construct the appropriate ICS29 Metadata type:
```
Comment (crodriguezvega) [inline docs/migrations/v3-to-v4.md:56]:
```suggestion
Similarly, if the application stack is configured to route through ICS29 fee middleware and a fee enabled channel is desired, construct the appropriate ICS29 `Metadata` type:
```

File: docs/migrations/v3-to-v4.md [hunk 4]
```diff
@@ -18,14 +18,72 @@ No genesis or in-place migrations required when upgrading from v1 or v2 of ibc-g
...
+### ICS27 - Interchain Accounts
+
+The `RegisterInterchainAccount` API has been modified to include an additional `version` argument. This change has been made in order to support ICS29 fee middelware, for relayer incentivization of ICS27 packets.
+Consumers of the `RegisterInterchainAccount` are now expected to build the appropriate JSON encoded version string themselves and pass it accordingly. 
+This should be constructed within the interchain accounts authentication module which leverages the APIs exposed via the interchain accounts `controllerKeeper`. 
+
+The following code snippet illustrates how to construct an appropriate interchain accounts Metadata and encode it as a JSON bytestring:
+
+```go
+icaMetadata := icatypes.Metadata{
```
Comment (crodriguezvega) [inline docs/migrations/v3-to-v4.md:38]: Since we have a constructor function (`NewMetadata`) available, shouldn't we encourage developers to use it?

**Review by crodriguezvega — APPROVED** [review]
Fantastic work, @damiannolan!

File: docs/migrations/v3-to-v4.md (hunk 4 above)
Comment (crodriguezvega) [reply docs/migrations/v3-to-v4.md:38]: Your arguments sound reasonable to me, so I am fine leaving as it is. :)

File: docs/migrations/v3-to-v4.md [hunk 5]
```diff
@@ -18,14 +18,72 @@ No genesis or in-place migrations required when upgrading from v1 or v2 of ibc-g
...
-### IS04 - Channel 
+### ICS04 - Channel 
 
 The `WriteAcknowledgement` API now takes the `exported.Acknowledgement` type instead of passing in the acknowledgement byte array directly. 
 This is an API breaking change and as such IBC application developers will have to update any calls to `WriteAcknowledgement`. 
 
 The `OnChanOpenInit` application callback has been modified.
 The return signature now includes the application version as detailed in the latest IBC [spec changes](https://github.com/cosmos/ibc/pull/629).
 
+### ICS27 - Interchain Accounts
```
Comment (AdityaSripal) [inline docs/migrations/v3-to-v4.md:29]: Nice docs!

**Review by AdityaSripal — APPROVED** [review]

**Pull request merged**